cluster/job_once_scheduler.go

```diff
@@ -0,0 +1,236 @@
+	if callback == nil {
+		return errors.New("callback cannot be nil")
+	}
+
+	s.storedCallback.mu.Lock()
+	defer s.storedCallback.mu.Unlock()
+
+	s.storedCallback.callback = callback
+	return nil
+}
+
+// ListScheduledJobs returns a list of the jobs in the db that have been scheduled. There is no
+// guarantee that list is accurate by the time the caller reads the list. E.g., the jobs in the list
+// may have been run, closed, or new jobs may have scheduled.
+func (s *JobOnceScheduler) ListScheduledJobs() ([]JobOnceMetadata, error) {
+	var ret []JobOnceMetadata
+	for i := 0; ; i++ {
+		keys, err := s.pluginAPI.KVList(i, keysPerPage)
+		if err != nil {
+			return nil, errors.Wrap(err, "error getting KVList")
+		}
+		for _, k := range keys {
+			if strings.HasPrefix(k, oncePrefix) {
+				metadata, err := readMetadata(s.pluginAPI, k[len(oncePrefix):])
+				if err != nil {
+					s.pluginAPI.LogError(errors.Wrap(err, "could not retrieve data from plugin kvstore for key: "+k).Error())
+					continue
+				}
+				if metadata == nil {
+					continue
+				}
+
+				ret = append(ret, *metadata)
+			}
+		}
+
+		if len(keys) < keysPerPage {
+			break
+		}
+	}
+
+	return ret, nil
+}
+
+// ScheduleOnce creates a scheduled job that will run once. When the clock reaches runAt, the
+// callback will be called with key as the argument.
+//
+// If the job key already exists in the db, this will return an error. To reschedule a job, first
+// close the original then schedule it again.
+func (s *JobOnceScheduler) ScheduleOnce(key string, runAt time.Time) (*JobOnce, error) {
+	s.startedMu.RLock()
+	defer s.startedMu.RUnlock()
+	if !s.started {
+		return nil, errors.New("start the scheduler before adding jobs")
+	}
+
+	job, err := newJobOnce(s.pluginAPI, key, runAt, s.storedCallback, s.activeJobs)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not create new job")
+	}
+
+	if err = job.saveMetadata(); err != nil {
+		return nil, errors.Wrap(err, "could not save job metadata")
+	}
+
+	s.runAndTrack(job)
+
+	return job, nil
+}
+
+// Close closes a job by its key. This is useful if the plugin lost the original *JobOnce, or
+// is stopping a job found in ListScheduledJobs().
+func (s *JobOnceScheduler) Close(key string) {
+	// using an anonymous function because job.Close() below needs access to the activeJobs mutex
+	job := func() *JobOnce {
+		s.activeJobs.mu.RLock()
+		defer s.activeJobs.mu.RUnlock()
+		j, ok := s.activeJobs.jobs[key]
+		if ok {
+			return j
+		}
+
+		// Job wasn't active, so no need to call Close (which shuts down the goroutine).
+		// There's a condition where another server in the cluster started the job, and the
+		// current server hasn't polled for it yet. To solve that case, delete it from the db.
+		mutex, err := NewMutex(s.pluginAPI, key)
+		if err != nil {
+			s.pluginAPI.LogError(errors.Wrap(err, "failed to create job mutex in Close for key: "+key).Error())
+		}
+		mutex.Lock()
+		defer mutex.Unlock()
```

Comment:
I'm not sure I understand why we need another mutex here.

Reply:
Because the job could exist on another server(s), so we need to lock before making any db change for that key.

Reply:
Aaaaah, yeah, it makes sense now, thanks!